rfcs/2020-09-29-4155-aws-s3-source.md

```diff
@@ -0,0 +1,187 @@
+## RFC #4155 - 2020-09-29 - AWS S3 source
+
+This RFC describes a new `aws_s3` source for ingesting events stored as objects
+in S3 via SQS bucket notifications. This will permit, for example, ingesting
+CloudTrail events, VPC flow logs, and ELB logs which can be written to S3, as
+well as custom logging pipelines writing there.
+
+## Scope
+
+This RFC will cover the implementation of an `aws_s3` source that will rely on
+SQS for bucket notifications.
+
+It will not cover the [proposed polling
+strategy](https://github.com/timberio/vector/issues/1017#issuecomment-694287111),
+but will leave space for it to be added later in a backwards compatible fashion.
+
+It will not cover parsing of events within the objects. Vector's current
+approach for this is to delegate to transforms. For example, it is likely we'll
+want to add a transform to extract CloudTrail events stored as objects in S3.
+
+## Motivation
+
+There are some types of logs that AWS can write to S3. For example:
+
+* CloudTrail logs (can also go to CloudWatch Logs but with a size limitation)
+* AWS load balancer access logs (only S3)
+* VPC flow logs (can also go to CloudWatch logs)
```

Comment:
Do we have a proposed approach for parsing the LB/VPC events? Can our existing transforms work here or do we need a transform?

Reply:
The VPC and ELB logs are whitespace separated values with one value per line.  So they can be easily tackled with the tokenizer transform.  
I think it might be nice to support the default configuration of the logs so that the user doesn't have to add the tokenizer transform step in the default case. 
https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html#flow-logs-default

Reply:
Agreed, I think the `tokenizer` transform will be sufficient for access logs and VPC flow logs. We'll probably want to provide some example configurations that can be used out-of-the-box with these formats.

For example, the VPC flow logs config, using `tokenizer` could look like:

<version> <account-id> <interface-id> <srcaddr> <dstaddr> <srcport> <dstport> <protocol> <packets> <bytes> <start> <end> <action> <log-status>
```toml
[transforms.my_transform_id]
  # General
  type = "tokenizer" # required
  inputs = ["my-source-or-transform-id"] # required
  drop_field = true # optional, default
  field = "message" # optional, default
  field_names = ["version", "account_id", "interface_id", "srcaddr", "dstaddr", "srcport", "dstport", "protocol", "packets", "bytes", "start", "end", "action", "log_status"] # required

  # Types
  types.version = "string"
  types.account_id = "string"
  types.interface_id = "string"
  types.srcaddr = "string"
  types.dstaddr = "string"
  types.srcport = "int"
  types.dstport = "int"
  types.protocol = "string"
  types.packets = "int"
  types.bytes = "int"
  types.start = "timestamp|%s"
  types.end = "timestamp|%s"
  types.action = "string"
  types.log_status = "string"
```

Again, this is probably useful to think about for configuration macros. cc/ @binarylogic

Reply:
I've also found the `regex_parser` to work well for fine-grained ELB log field parsing. Here's a sample ELB parsing config I've used previously:
```
[transforms.elasticloadbalancing_fields_parsed]
  type = "regex_parser"
  inputs = ["..."]
  regex = '^(?P<type>[\w]+) (?P<timestamp>[\w:.-]+) (?P<elb>[^\s]+) (?P<client_host>[\d.:-]+) (?P<target_host>[\d.:-]+) (?P<request_processing_time>[\d.-]+) (?P<target_processing_time>[\d.-]+) (?P<response_processing_time>[\d.-]+) (?P<elb_status_code>[\d-]+) (?P<target_status_code>[\d-]+) (?P<received_bytes>[\d-]+) (?P<sent_bytes>[\d-]+) "(?P<request_method>[\w-]+) (?P<request_url>[^\s]+) (?P<request_protocol>[^"\s]+)" "(?P<user_agent>[^"]+)" (?P<ssl_cipher>[^\s]+) (?P<ssl_protocol>[^\s]+) (?P<target_group_arn>[\w.:/-]+) "(?P<trace_id>[^\s"]+)" "(?P<domain_name>[^\s"]+)" "(?P<chosen_cert_arn>[\w:./-]+)" (?P<matched_rule_priority>[\d-]+) (?P<request_creation_time>[\w.:-]+) "(?P<actions_executed>[\w,-]+)" "(?P<redirect_url>[^"]+)" "(?P<error_reason>[^"]+)"'
  field = "message"
  drop_failed = false

  types.received_bytes = "int"
  types.request_processing_time = "float"
  types.sent_bytes = "int"
  types.target_processing_time = "float"
  types.response_processing_time = "float"

[transforms.elasticloadbalancing_url_parsed]
  type = "regex_parser"
  inputs = ["elasticloadbalancing_fields_parsed"]
  regex = '^(?P<url_scheme>[\w]+)://(?P<url_hostname>[^\s:/?#]+)(?::(?P<request_port>[\d-]+))?-?(?:/(?P<url_path>[^\s?#]*))?(?P<request_url_query>\?[^\s#]+)?'
  field = "request_url"
  drop_failed = false
```


Which turns:
```
h2 2018-07-02T22:23:00.186641Z app/my-loadbalancer/50dc6c495c0c9188 10.0.1.252:48160 10.0.0.66:9000 0.000 0.002 0.000 200 200 5 257 "GET https://10.0.2.105:773/ HTTP/2.0" "curl/7.46.0" ECDHE-RSA-AES128-GCM-SHA256 TLSv1.2 arn:aws:elasticloadbalancing:us-east-2:123456789012:targetgroup/my-targets/73e2d6bc24d8a067 "Root=1-58337327-72bd00b0343d75b906739c42" "-" "-" 1 2018-07-02T22:22:48.364000Z "redirect" "https://example.com:80/" "-" 10.0.0.66:9000 200 "-" "-"
```
into:
```
{
  "actions_executed": "redirect",
  "chosen_cert_arn": "-",
  "client_host": "10.0.1.252:48160",
  "domain_name": "-",
  "elb": "app/my-loadbalancer/50dc6c495c0c9188",
  "elb_status_code": "200",
  "error_reason": "-",
  "matched_rule_priority": "1",
  "received_bytes": 5,
  "redirect_url": "https://example.com:80/",
  "request_creation_time": "2018-07-02T22:22:48.364000Z",
  "request_method": "GET",
  "request_port": "773",
  "request_processing_time": 0,
  "request_protocol": "HTTP/2.0",
  "response_processing_time": 0,
  "sent_bytes": 257,
  "ssl_cipher": "ECDHE-RSA-AES128-GCM-SHA256",
  "ssl_protocol": "TLSv1.2",
  "target_group_arn": "arn:aws:elasticloadbalancing:us-east-2:123456789012:targetgroup/my-targets/73e2d6bc24d8a067",
  "target_host": "10.0.0.66:9000",
  "target_processing_time": 0.002,
  "target_status_code": "200",
  "timestamp": "2018-07-02T22:23:00.186641Z",
  "trace_id": "Root=1-58337327-72bd00b0343d75b906739c42",
  "type": "h2",
  "url_hostname": "10.0.2.105",
  "url_path": "",
  "url_scheme": "https",
  "user_agent": "curl/7.46.0"
}
```

Reply:
Nice, thanks for that @zcapper.